### PR TITLE
[Flang][MLIR][OpenMP] Align map clause generation and fix issue with non-shared allocations for assumed shape/size descriptor types

### PR DESCRIPTION
**Pull request by agozillon**
This PR aims to unify the map argument generation behavior across both the implicit capture (captured in a target region) and the explicit capture (process map), currently the varPtr field of the MapInfo for the same variable will be different depending on how it's captured. This PR tries to align that across the generations of MapInfoOp in the OpenMP lowering.

Currently, I have opted to utilise the rawInput (input memref to a HLFIR DeclareInfoOp) as opposed to the addr field which includes more information. The side affect of this is that we have to deal with BoxTypes less often, which will result in simpler maps in these cases. The negative side affect of this is that we don't have access to the bounds information through the resulting value, however, I believe the bounds information we require in our case is still appropriately stored in the map bounds, and this seems to be the case from testing so far.

The other fix is for cases where we end up with a BoxType argument into a function (certain assumed shape and sizes cases do this) that has no fir.ref wrapping it. As we need the Box to be a reference type to actually utilise the operation to access the base address stored inside and create the correct mappings we currently generate an intermediate allocation in these cases, and then store into it, and utilise this as the map argument, as opposed to the original.

However, as we were not sharing the same intermediate allocation across all of the maps for a variable, this resulted in errors in certain cases when detatching/attatching the data e.g. via enter and exit.  This PR adjusts this for cases

Currently we only maintain tracking of all intermediate allocations for the current function scope, as opposed to module. Primarily as the only case I am aware of that this is required is in cases where we pass certain types of arguments to functions (so I opted to minimize the overhead of the pass for now). It could likely be extended to module scope if required if we find other cases where it's applicable and causing issues.

### CODE REVIEW

**Comment by agozillon** [comment]
I plan on making an upstream PR for this as well soon and I'll ask Jean Perrier to have a look over it, primarily for the usage of the rawInput, to see if there's any side affects to look out for in the cases where it provides a result from a hlfir declare op! 

However, until then I would be interested in applying this to ATD to see if it improves or decreases our nightly builds pass rates. I would imagine improve by a small amount, but I'd prefer to test it on our nightlies before landing it upstream where it will be more of a pain to fix when it's in!

**Review by ronlieb — APPROVED** [review]
LGTM from a build/test perspective.

**Comment by agozillon** [comment]
@ronlieb Thank you very much! I'll land it on Monday then provided no one has any further comments then, so as not to have the nightly results bother or confuse anyone over the weekend if it does do anything adverse I wasn't expecting...

File: flang/lib/Lower/OpenMP/OpenMP.cpp
Comment (agozillon) [reply flang/lib/Lower/OpenMP/OpenMP.cpp:2023]: This change already exists, the previous code was doing the same thing, it's just more explicit in the above case. However, it's likely possible to omit this check + early return as I would imagine it's already being caught by the addition on line 2038. A similar early return is done later, in the body generation when we're handling symbol re-bindings to target arguments. But in either case without these in place, it would be a compiler ICE as opposed to an IR change in cases where derived type members are implicitly captured, as there symbol is the parent and they themselves do not have symbols currently (from my understanding at least).

**Review by skatrak — APPROVED** [review]
I'm not too familiar with the code being changed, but changes make sense to me. I just have minimal comment-related nits. LGTM, but please wait for approval by @ergawy before merging. Thank you!

**Comment by agozillon** [comment]
Updated the PR to hopefully cover all comments!

**Comment by agozillon** [comment]
Thank you everyone, going to run a final set of tests just to check I didn't break anything with the NIT adjustments then I'll merge the PR.

**Pull request merged**
